crates/cargo-contract/src/cmd/extrinsics/balance.rs

```diff
@@ -0,0 +1,670 @@
+                    return Err(anyhow!(
+                        "Given precision of a Balance value is higher than allowed"
+                    ))
+                }
+                let balance: Balance = (den_balance.value * multiple).try_into()?;
+                Ok(balance)
+            }
+        }
+    }
+
+    /// # Summary
+    /// Display token units in a denominated format.
+    ///
+    /// I takes `value` of `Into<u128>` and [TokenMetadata]
+    /// and calculates the value in an denominated format
+    /// by manipulating the token_decimals.
+    ///
+    /// If the number is divisible by 10^(`token_decimals` + `unit_zeros`),
+    /// It sets the [UnitPrefix] and divides the `value` into `Decimal`
+    ///
+    /// If no [TokenMetadata] was present, than that means
+    /// that [Balance] is to be displayed in *normal* format
+    /// and `BalanceVariant::Default` is returned
+    ///
+    /// # Examples
+    /// ```rust
+    ///  use anyhow::{Result, Ok};
+    ///  let decimals = 10;
+    ///  let tm = TokenMetadata {
+    ///        token_decimals: decimals,
+    ///        symbol: String::from("DOT"),
+    /// };
+    /// let sample_den_balance = BalanceVariant::Denominated(DenominatedBalance {
+    ///     value: Decimal::new(5005, 1),
+    ///     unit: UnitPrefix::Mega,
+    ///     symbol: String::from("DOT")
+    /// });
+    /// let balance: Balance = 5_005_000_000_000_000_000;
+    /// let den_balance = BalanceVariant::from(balance, Some(tm));
+    /// assert_eq!(Ok(sample_den_balance), Ok(den_balance));
+    /// ```
+    pub fn from<T: Into<u128>>(
+        value: T,
+        token_metadata: Option<&TokenMetadata>,
+    ) -> Result<Self> {
+        let n: u128 = value.into();
+
+        if let Some(token_metadata) = token_metadata {
+            if n == 0 {
+                return Ok(BalanceVariant::Denominated(DenominatedBalance {
+                    value: Decimal::ZERO,
+                    unit: UnitPrefix::One,
+                    symbol: token_metadata.symbol.clone(),
+                }))
+            }
+
+            let number_of_digits = n.to_string().len();
+
+            let giga_units_zeros = token_metadata.token_decimals + 9;
+            let mega_units_zeros = token_metadata.token_decimals + 6;
+            let kilo_units_zeros = token_metadata.token_decimals + 3;
+            let one_unit_zeros = token_metadata.token_decimals;
+            let milli_units_zeros = token_metadata.token_decimals.checked_sub(3);
+            let micro_units_zeros = token_metadata.token_decimals.checked_sub(6);
+            let nano_units_zeros = token_metadata.token_decimals.checked_sub(9);
+
+            let multiple: Decimal;
+            let unit: UnitPrefix;
+            if (giga_units_zeros + 1..).contains(&number_of_digits) {
+                multiple = Decimal::from_str_exact(&format!(
+                    "1{}",
+                    "0".repeat(giga_units_zeros)
+                ))?;
```

Comment:
This line is repeated in each conditional branch, prefer to return the no of zeros from the conditional and then do the `Decimal::from_str_exact` once at the end.